framework/keyring-interface.md

```diff
@@ -75,66 +86,65 @@ perform the [Encrypt Data Key](#encrypt-data-key) behavior.
 
 To perform this behavior, the keyring creates one or more [encrypted data keys](#structures.md#encrypted-data-key)
-using the plaintext data key from the [encryption materials](#structures.md#encryption-materials) as input,
-and appends the [encrypted data keys](#structures.md#encrypted-data-key) to the encrypted data key list
-in the [encryption materials](#structures.md#encryption-materials).
+using the plaintext data key as input,
+and includes the [encrypted data keys](#structures.md#encrypted-data-key)
+in the [data key materials](#structures.md#data-key-materials) result.
 
 The [encrypted data keys](#structures.md#encrypted-data-key) produced by this keyring MUST
 have [ciphertexts](#structures.md#ciphertext) that can be decrypted to the plaintext data key in the
-[encryption materials](#structures.md#encryption-materials).
+[data key materials](#structures.md#data-key-materials).
 
-If OnEncrypt updates the [encryption materials](#data-structure.md#encryption-materials) with at least
-one new [encrypted data key](#structures.md#encrypted-data-key),
+If OnEncrypt outputs [data key materials](#structures.md#data-key-materials) with at least
+one [encrypted data key](#structures.md#encrypted-data-key),
 the [keyring trace](#structures.md#keyring-trace) returned by OnEncrypt MUST include at least one trace
 associated with this keyring that has the [ENCRYPTED DATA KEY](#structures.md#encrypted-data-key) flag.
 Note that this trace MAY include more than one flag,
 for example the [SIGNED ENCRYPTION CONTEXT flag](#structures.md#signed-encryption-context).
 
 ### OnDecrypt
 
-This interface takes [decryption materials](#structures.md#decryption-materials) and
-a list of [encrypted data keys](#structures.md#encrypted-data-key) as input and
-MAY modify it with the following behavior:
+The following inputs are REQUIRED:
+
+- [Algorithm Suite](#algorithm-suites.md)
+- [Encryption Context](#structures.md#encryption-context)
+- [Encrypted Data Keys](#structures.md#encrypted-data-keys)
+
+This interface MAY perform the following behavior:
 
 - [Decrypt data key](#decrypt-data-key)
 
-If this keyring attempted the above behavior, and succeeded, it MUST output the modified [decryption materials](#structures.md#decryption-materials).
+If this keyring attempted the above behavior, and succeeded, it MUST output the resulting [data key materials](#structures.md#data-key-materials). The [algorithm suite](#algorithm-suites.md) in the result must match the input [algorithm suite](#algorithm-suites.md), and the [encrypted data keys](#structures.md#encrypted-data-keys)
+list must be empty.
 
-If the keyring did not attempt the above behavior,
-the keyring MUST output the [decryption materials](#structures.md#decryption-materials) unmodified.
+If the keyring did not attempt the above behavior, it MUST produce no output.
 
 #### Decrypt Data Key
 
-If the encryption materials do contain a plaintext data key, OnDecrypt MUST NOT decrypt a data key.
```

Comment:
Actually the error is that I didn't remove the plaintext data key from the decryption materials - that was an "output parameter" for `Keyring.OnDecrypt` so it should no longer be part of the input, which means this statement no longer makes sense.

Reply:
Ok, so what that does is move the responsibility of "deciding" whether to actually do OnDecrypt (in the case that the caller already has a plaintextDataKey) on the caller instead of the keyring itself.

This should be allowed, and is what the C implementation already does, but I'm unsure how we want to capture this in the spec, or if leaving any description out entirely is the right answer.